src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.Validation.cs

```diff
@@ -119,20 +121,30 @@ internal static unsafe partial class Utf8Utility
                         byte* pInputBufferFinalPosAtWhichCanSafelyLoop = pFinalPosWhereCanReadDWordFromInputBuffer - 3 * sizeof(uint); // can safely read 4 DWORDs here
                         uint mask;
 
+                        Vector128<byte> initialMask = Vector128.Create((ushort)0x1001).AsByte();
+                        Vector128<byte> mostSignficantBitMask = Vector128.Create((byte)0x80).AsByte();
+
                         do
                         {
-                            if (Sse2.IsSupported)
+                            // pInputBuffer is 32-bit aligned but not necessary 128-bit aligned, so we're
+                            // going to perform an unaligned load. We don't necessarily care about aligning
+                            // this because we pessimistically assume we'll encounter non-ASCII data at some
+                            // point in the not-too-distant future (otherwise we would've stayed entirely
+                            // within the all-ASCII vectorized code at the entry to this method).
+                            if (AdvSimd.Arm64.IsSupported && BitConverter.IsLittleEndian)
                             {
-                                // pInputBuffer is 32-bit aligned but not necessary 128-bit aligned, so we're
-                                // going to perform an unaligned load. We don't necessarily care about aligning
-                                // this because we pessimistically assume we'll encounter non-ASCII data at some
-                                // point in the not-too-distant future (otherwise we would've stayed entirely
-                                // within the all-ASCII vectorized code at the entry to this method).
-
-                                mask = (uint)Sse2.MoveMask(Sse2.LoadVector128((byte*)pInputBuffer));
+                                mask = (uint)Arm64MoveMask(AdvSimd.LoadVector128(pInputBuffer), initialMask, mostSignficantBitMask);
```

Comment:
>(uint) [](start = 39, length = 6)

you might not want to cast the `ulong` to `uint` specially because you are checking `mask == 0` . `4294967296` if type-casted to `uint` will return 0 but in reality its not.

Reply:
Also below use `BitOperations.TrailingZeroCount(ulong)` version.

---
In reply to: [453997457](https://github.com/dotnet/runtime/pull/38653#discussion_r453997457) [](ancestors = 453997457)

Reply:
Thanks. To address this, I decided to change the data type of `mask` to `ulong` in my latest commit. Since `Sse2.MoveMask` was being casted from `int` to `uint`, I think a similar problem could happen there since some data would be lost. Is it ok to cast the existing code to a larger integer type? That ensured I could still reuse the code.

Reply:
This problem can't happen in `MoveMask` because you are operating on 16 lanes and in case they all have MSB set, then maximum possible value you can get back from `MoveMask()` is `65535` which is within range of `uint`. 

---
In reply to: [454083836](https://github.com/dotnet/runtime/pull/38653#discussion_r454083836) [](ancestors = 454083836)

Reply:
Ah yes, thanks for the correction. uint [holds more data](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/integral-numeric-types).